test/sm2crypttest.c

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <openssl/bio.h>
+#include <openssl/evp.h>
+#include <openssl/bn.h>
+#include <openssl/crypto.h>
+# include <openssl/err.h>
+# include <openssl/rand.h>
+#include "testutil.h"
+
+#ifndef OPENSSL_NO_SM2
+
+# include <openssl/sm2.h>
+
+static RAND_METHOD fake_rand;
+static const RAND_METHOD *saved_rand;
+
+uint8_t* fake_rand_bytes = NULL;
+size_t fake_rand_bytes_offset = 0;
+
+static int get_faked_bytes(unsigned char *buf, int num)
+{
+    int i;
+
+    if (fake_rand_bytes == NULL)
+        return saved_rand->bytes(buf, num);
+
+    for(i = 0; i != num; ++i)
+       buf[i] = fake_rand_bytes[fake_rand_bytes_offset + i];
+    fake_rand_bytes_offset += num;
+    return 1;
+}
+
+
+static int start_fake_rand(const char* hex_bytes)
+{
+    /* save old rand method */
+    if (!TEST_ptr(saved_rand = RAND_get_rand_method()))
+        return 0;
+
+    fake_rand = *saved_rand;
+    /* use own random function */
+    fake_rand.bytes = get_faked_bytes;
+
+    fake_rand_bytes = OPENSSL_hexstr2buf(hex_bytes, NULL);
+    fake_rand_bytes_offset = 0;
+
+    /* set new RAND_METHOD */
+    if (!TEST_true(RAND_set_rand_method(&fake_rand)))
+        return 0;
+    return 1;
+}
+
+static int restore_rand(void) {
+    OPENSSL_free(fake_rand_bytes);
+    fake_rand_bytes = NULL;
+    fake_rand_bytes_offset = 0;
+    if (!TEST_true(RAND_set_rand_method(saved_rand)))
+        return 0;
+    return 1;
+}
+
+static EC_GROUP *create_EC_group(
+   const char *p_hex, const char *a_hex, const char *b_hex,
+   const char *x_hex, const char *y_hex,
+   const char *order_hex, const char *cof_hex)
+   {
+   BIGNUM* p = NULL;
+   BIGNUM* a = NULL;
+   BIGNUM* b = NULL;
+   BIGNUM* g_x = NULL;
+   BIGNUM* g_y = NULL;
+   BIGNUM* order = NULL;
+   BIGNUM* cof = NULL;
+   EC_POINT* generator = NULL;
+   EC_GROUP *group = NULL;
+
+   BN_hex2bn(&p, p_hex);
+   BN_hex2bn(&a, a_hex);
+   BN_hex2bn(&b, b_hex);
+
+   group = EC_GROUP_new_curve_GFp(p, a, b, NULL);
+   BN_free(p);
+   BN_free(a);
+   BN_free(b);
+
+   if(group == NULL)
+      return NULL;
+
+   generator = EC_POINT_new(group);
+   if(generator == NULL)
+      return NULL;
+
+   BN_hex2bn(&g_x, x_hex);
+   BN_hex2bn(&g_y, y_hex);
+
+   if(EC_POINT_set_affine_coordinates_GFp(group, generator, g_x, g_y, NULL) == 0)
+      return NULL;
+
+   BN_free(g_x);
+   BN_free(g_y);
+
+   BN_hex2bn(&order, order_hex);
+   BN_hex2bn(&cof, cof_hex);
+
+   if(EC_GROUP_set_generator(group, generator, order, cof) == 0)
+      return NULL;
+
+   EC_POINT_free(generator);
+   BN_free(order);
+   BN_free(cof);
+
+   return group;
+   }
```

Comment:
Style: this whole function need to be fixed; outermost `{` and `}` in column 1, function parameters aligned within the parenthesis, 4 space indent instead of 3, and `if(` changed to `if (`